test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Audience.cs

```diff
@@ -0,0 +1,241 @@
+
+            Saml2SecurityTokenHandler saml2TokenHandler = new Saml2SecurityTokenHandler();
+
+            var saml2Token = CreateToken(theoryData.TokenAudience!, theoryData.Saml2Condition!);
+
+            var tokenValidationParameters = CreateTokenValidationParameters(
+                theoryData.TVPAudiences,
+                saml2Token,
+                theoryData.NullTokenValidationParameters,
+                theoryData.IgnoreTrailingSlashWhenValidatingAudience);
+
+            // Validate the token using TokenValidationParameters
+            TokenValidationResult tokenValidationResult =
+                await saml2TokenHandler.ValidateTokenAsync(saml2Token.Assertion.CanonicalString, tokenValidationParameters);
+
+            // Validate the token using ValidationParameters.
+            ValidationResult<ValidatedToken> validationResult =
+                await saml2TokenHandler.ValidateTokenAsync(
+                    saml2Token,
+                    theoryData.ValidationParameters!,
+                    theoryData.CallContext,
+                    CancellationToken.None);
+
+            // Ensure the validity of the results match the expected result.
+            if (tokenValidationResult.IsValid != validationResult.IsSuccess)
+            {
+                context.AddDiff($"tokenValidationResult.IsValid != validationResult.IsSuccess");
+                theoryData.ExpectedExceptionValidationParameters!.ProcessException(validationResult.UnwrapError().GetException(), context);
+                theoryData.ExpectedException.ProcessException(tokenValidationResult.Exception, context);
+            }
+            else
+            {
+                if (tokenValidationResult.IsValid)
+                {
+                    // Verify that the validated tokens from both paths match.
+                    ValidatedToken validatedToken = validationResult.UnwrapResult();
+                    IdentityComparer.AreEqual(validatedToken.SecurityToken, tokenValidationResult.SecurityToken, context);
+                }
+                else
+                {
+                    // Verify the exception provided by both paths match.
+                    var tokenValidationResultException = tokenValidationResult.Exception;
+                    theoryData.ExpectedException.ProcessException(tokenValidationResult.Exception, context);
+                    var validationResultException = validationResult.UnwrapError().GetException();
+                    theoryData.ExpectedExceptionValidationParameters!.ProcessException(validationResult.UnwrapError().GetException(), context);
+                }
+
+                TestUtilities.AssertFailIfErrors(context);
+            }
+        }
+
+        public static TheoryData<ValidateTokenAsyncAudienceTheoryData> ValidateTokenAsync_Audience_TestCases
+        {
+            get
+            {
+                return new TheoryData<ValidateTokenAsyncAudienceTheoryData>
```

Comment:
I know it seems cleaner to write and VS will suggest this: 
```cs
return new TheoryData<ValidateTokenAsyncAudienceTheoryData>
{
   new ValidateTokenAsyncAudienceTheoryData();
}
```
than
```cs
var theoryData = new TheoryData<ValidateTokenAsyncAudienceTheoryData>();

theoryData.Add (new ValidateTokenAsyncAudienceTheoryData());

return theoryData;
```
However it is difficult to put a breakpoint on a single test case which sometimes is helpful.
For this reason i tend to use the .Add() model.

Reply:
That's right, I haven't fully gotten used to debugging with breakspoints in theories, I updated them. Thanks Brent!